### PR TITLE
Replace custom Cow with actual Cow

### PR DESCRIPTION
**Pull request by Kijewski**
[Cow](https://flickr.com/photos/jeena/9342245281/)

### CODE REVIEW

File: askama_shared/src/lib.rs [hunk 1]
```diff
@@ -294,29 +293,26 @@ static DEFAULT_ESCAPERS: &[(&[&str], &str)] = &[
...
-        }
+        fmt::Display::fmt(&self.0, fmt)
     }
 }
 
 impl From<&'static str> for CompileError {
+    #[inline]
     fn from(s: &'static str) -> Self {
-        CompileError::Static(s)
+        CompileError(s.into())
```
Comment (vallentin) [inline askama_shared/src/lib.rs:308]:
```suggestion
        Self(s.into())
```

Odd, that this case doesn't trigger the lint. Anyways, might be good to preemptively fix that as well :)

File: askama_shared/src/lib.rs [hunk 2]
```diff
@@ -294,29 +293,26 @@ static DEFAULT_ESCAPERS: &[(&[&str], &str)] = &[
...
     }
 }
 
 impl From<String> for CompileError {
+    #[inline]
     fn from(s: String) -> Self {
-        CompileError::String(s)
+        CompileError(s.into())
```
Comment (vallentin) [inline askama_shared/src/lib.rs:315]:
```suggestion
        Self(s.into())
```

(Same as before)

**Review by djc — COMMENTED** [review]
Thanks!

File: askama_shared/src/lib.rs [hunk 3]
```diff
@@ -294,29 +293,26 @@ static DEFAULT_ESCAPERS: &[(&[&str], &str)] = &[
...
+pub struct CompileError(Cow<'static, str>);
 
 impl fmt::Display for CompileError {
+    #[inline]
     fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            CompileError::Static(s) => write!(fmt, "{}", s),
-            CompileError::String(s) => write!(fmt, "{}", s),
-        }
+        fmt::Display::fmt(&self.0, fmt)
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply askama_shared/src/lib.rs]: Makes sense, it's one less indirection.

File: askama_shared/src/generator.rs [hunk 4]
```diff
@@ -665,27 +665,25 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
...
-            (
-                mctx.macros.get(name).ok_or_else(|| {
-                    CompileError::String(format!("macro '{}' not found in scope '{}'", name, s))
-                })?,
-                mctx,
-            )
-        } else {
-            (
-                ctx.macros
+        let (def, own_ctx) =
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply askama_shared/src/generator.rs]: It is. Done.

**Review by djc — APPROVED** [review]

**Pull request merged**